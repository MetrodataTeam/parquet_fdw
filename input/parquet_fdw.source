SET client_min_messages = WARNING;

CREATE EXTENSION parquet_fdw;

DROP ROLE IF EXISTS regress_parquet_fdw;
CREATE ROLE regress_parquet_fdw LOGIN SUPERUSER;

SET ROLE regress_parquet_fdw;
CREATE SERVER parquet_srv FOREIGN DATA WRAPPER parquet_fdw;
CREATE USER MAPPING FOR regress_parquet_fdw SERVER parquet_srv;

SET ROLE regress_parquet_fdw;

CREATE FOREIGN TABLE example (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example.parquet', sorted 'one');

SELECT * FROM example;
-- sorting
EXPLAIN (COSTS OFF) SELECT * FROM example ORDER BY one;
EXPLAIN (COSTS OFF) SELECT * FROM example ORDER BY three;

-- invalid options
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv;
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename 'nonexistent.parquet', some_option '123');
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example.parquet', some_option '123');
