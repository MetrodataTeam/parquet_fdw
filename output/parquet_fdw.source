SET datestyle = 'ISO';
SET client_min_messages = WARNING;
CREATE EXTENSION parquet_fdw;
DROP ROLE IF EXISTS regress_parquet_fdw;
CREATE ROLE regress_parquet_fdw LOGIN SUPERUSER;
SET ROLE regress_parquet_fdw;
CREATE SERVER parquet_srv FOREIGN DATA WRAPPER parquet_fdw;
CREATE USER MAPPING FOR regress_parquet_fdw SERVER parquet_srv;
SET ROLE regress_parquet_fdw;
CREATE FOREIGN TABLE example1 (
    one     INT8,
    two     INT8[],
    three   TEXT,
    four    TIMESTAMP,
    five    DATE,
    six     BOOL)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example1.parquet', sorted 'one');
SELECT * FROM example1;
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   1 | {1,2,3}    | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
   3 | {7,8,9}    | baz   | 2018-01-03 00:00:00 | 2018-01-03 | t
   4 | {10,11,12} | uno   | 2018-01-04 00:00:00 | 2018-01-04 | f
   5 | {13,14,15} | dos   | 2018-01-05 00:00:00 | 2018-01-05 | f
   6 | {16,17,18} | tres  | 2018-01-06 00:00:00 | 2018-01-06 | f
(6 rows)

-- no explicit columns mentions
SELECT 1 as x FROM example1;
 x 
---
 1
 1
 1
 1
 1
 1
(6 rows)

SELECT count(*) as count FROM example1;
 count 
-------
     6
(1 row)

-- sorting
EXPLAIN (COSTS OFF) SELECT * FROM example1 ORDER BY one;
        QUERY PLAN        
--------------------------
 Foreign Scan on example1
   Reader: Single File
   Row groups: 1, 2
(3 rows)

EXPLAIN (COSTS OFF) SELECT * FROM example1 ORDER BY three;
           QUERY PLAN           
--------------------------------
 Sort
   Sort Key: three
   ->  Foreign Scan on example1
         Reader: Single File
         Row groups: 1, 2
(5 rows)

-- filtering
SET client_min_messages = DEBUG1;
SELECT * FROM example1 WHERE one < 1;
DEBUG:  parquet_fdw: skip rowgroup 1
DEBUG:  parquet_fdw: skip rowgroup 2
 one | two | three | four | five | six 
-----+-----+-------+------+------+-----
(0 rows)

SELECT * FROM example1 WHERE one <= 1;
DEBUG:  parquet_fdw: skip rowgroup 2
 one |   two   | three |        four         |    five    | six 
-----+---------+-------+---------------------+------------+-----
   1 | {1,2,3} | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
(1 row)

SELECT * FROM example1 WHERE one > 6;
DEBUG:  parquet_fdw: skip rowgroup 1
DEBUG:  parquet_fdw: skip rowgroup 2
 one | two | three | four | five | six 
-----+-----+-------+------+------+-----
(0 rows)

SELECT * FROM example1 WHERE one >= 6;
DEBUG:  parquet_fdw: skip rowgroup 1
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   6 | {16,17,18} | tres  | 2018-01-06 00:00:00 | 2018-01-06 | f
(1 row)

SELECT * FROM example1 WHERE one = 2;
DEBUG:  parquet_fdw: skip rowgroup 2
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
(1 row)

SELECT * FROM example1 WHERE one = 7;
DEBUG:  parquet_fdw: skip rowgroup 1
DEBUG:  parquet_fdw: skip rowgroup 2
 one | two | three | four | five | six 
-----+-----+-------+------+------+-----
(0 rows)

SELECT * FROM example1 WHERE six = true;
DEBUG:  parquet_fdw: skip rowgroup 2
 one |   two   | three |        four         |    five    | six 
-----+---------+-------+---------------------+------------+-----
   1 | {1,2,3} | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
   3 | {7,8,9} | baz   | 2018-01-03 00:00:00 | 2018-01-03 | t
(2 rows)

SELECT * FROM example1 WHERE six = false;
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
   4 | {10,11,12} | uno   | 2018-01-04 00:00:00 | 2018-01-04 | f
   5 | {13,14,15} | dos   | 2018-01-05 00:00:00 | 2018-01-05 | f
   6 | {16,17,18} | tres  | 2018-01-06 00:00:00 | 2018-01-06 | f
(4 rows)

-- prepared statements
prepare prep(date) as select * from example1 where five < $1;
execute prep('2018-01-03');
DEBUG:  parquet_fdw: skip rowgroup 2
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   1 | {1,2,3}    | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
(2 rows)

execute prep('2018-01-01');
DEBUG:  parquet_fdw: skip rowgroup 1
DEBUG:  parquet_fdw: skip rowgroup 2
 one | two | three | four | five | six 
-----+-----+-------+------+------+-----
(0 rows)

-- invalid options
SET client_min_messages = WARNING;
CREATE FOREIGN TABLE example_fail (one INT8, two INT8[], three TEXT)
SERVER parquet_srv;
ERROR:  parquet_fdw: filename is required
CREATE FOREIGN TABLE example_fail (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename 'nonexistent.parquet', some_option '123');
ERROR:  parquet_fdw: No such file or directory
CREATE FOREIGN TABLE example_fail (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example1.parquet', some_option '123');
ERROR:  parquet_fdw: invalid option "some_option"
-- type mismatch
CREATE FOREIGN TABLE example_fail (one INT8[], two INT8, three TEXT)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example1.parquet', sorted 'one');
SELECT one FROM example_fail;
ERROR:  parquet_fdw: incompatible types in column "one"
HINT:  parquet column is of scalar type while postgres type is array
SELECT two FROM example_fail;
ERROR:  parquet_fdw: incompatible types in column "two"
HINT:  parquet column is of type list while postgres type is scalar
-- sequential multifile reader
CREATE FOREIGN TABLE example_seq (
    one     INT8,
    two     INT8[],
    three   TEXT,
    four    TIMESTAMP,
    five    DATE,
    six     BOOL)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example1.parquet @abs_srcdir@/data/example2.parquet');
EXPLAIN SELECT * FROM example_seq;
                           QUERY PLAN                            
-----------------------------------------------------------------
 Foreign Scan on example_seq  (cost=0.00..0.11 rows=11 width=85)
   Reader: Multifile
   Row groups: 
 example1.parquet: 1, 2
 example2.parquet: 1
(5 rows)

SELECT * FROM example_seq;
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   1 | {1,2,3}    | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
   3 | {7,8,9}    | baz   | 2018-01-03 00:00:00 | 2018-01-03 | t
   4 | {10,11,12} | uno   | 2018-01-04 00:00:00 | 2018-01-04 | f
   5 | {13,14,15} | dos   | 2018-01-05 00:00:00 | 2018-01-05 | f
   6 | {16,17,18} | tres  | 2018-01-06 00:00:00 | 2018-01-06 | f
   1 | {19,20}    | eins  | 2018-01-01 00:00:00 | 2018-01-01 | t
   3 | {21,22}    | zwei  | 2018-01-03 00:00:00 | 2018-01-03 | f
   5 | {23,24}    | drei  | 2018-01-05 00:00:00 | 2018-01-05 | t
   7 | {25,26}    | vier  | 2018-01-07 00:00:00 | 2018-01-07 | f
   9 | {27,28}    | fünf  | 2018-01-09 00:00:00 | 2018-01-09 | t
(11 rows)

-- multifile merge reader
CREATE FOREIGN TABLE example_merge (
    one     INT8,
    two     INT8[],
    three   TEXT,
    four    TIMESTAMP,
    five    DATE,
    six     BOOL)
SERVER parquet_srv
OPTIONS (filename '@abs_srcdir@/data/example1.parquet @abs_srcdir@/data/example2.parquet', sorted 'one');
EXPLAIN (COSTS OFF) SELECT * FROM example_merge ORDER BY one;
          QUERY PLAN           
-------------------------------
 Foreign Scan on example_merge
   Reader: Multifile Merge
   Row groups: 
 example1.parquet: 1, 2
 example2.parquet: 1
(5 rows)

SELECT * FROM example_merge ORDER BY one;
 one |    two     | three |        four         |    five    | six 
-----+------------+-------+---------------------+------------+-----
   1 | {19,20}    | eins  | 2018-01-01 00:00:00 | 2018-01-01 | t
   1 | {1,2,3}    | foo   | 2018-01-01 00:00:00 | 2018-01-01 | t
   2 | {NULL,5,6} | bar   | 2018-01-02 00:00:00 | 2018-01-02 | f
   3 | {21,22}    | zwei  | 2018-01-03 00:00:00 | 2018-01-03 | f
   3 | {7,8,9}    | baz   | 2018-01-03 00:00:00 | 2018-01-03 | t
   4 | {10,11,12} | uno   | 2018-01-04 00:00:00 | 2018-01-04 | f
   5 | {23,24}    | drei  | 2018-01-05 00:00:00 | 2018-01-05 | t
   5 | {13,14,15} | dos   | 2018-01-05 00:00:00 | 2018-01-05 | f
   6 | {16,17,18} | tres  | 2018-01-06 00:00:00 | 2018-01-06 | f
   7 | {25,26}    | vier  | 2018-01-07 00:00:00 | 2018-01-07 | f
   9 | {27,28}    | fünf  | 2018-01-09 00:00:00 | 2018-01-09 | t
(11 rows)

