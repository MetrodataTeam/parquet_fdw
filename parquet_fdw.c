#include "postgres.h"
#include "fmgr.h"

#include "access/reloptions.h"
#include "catalog/pg_foreign_table.h"
#include "foreign/fdwapi.h"
#include "optimizer/planmain.h"
#include "utils/elog.h"


PG_MODULE_MAGIC;


/* FDW routines */
extern void parquetGetForeignRelSize(PlannerInfo *root,
					  RelOptInfo *baserel,
					  Oid foreigntableid);
extern void parquetGetForeignPaths(PlannerInfo *root,
					RelOptInfo *baserel,
					Oid foreigntableid);
extern ForeignScan *parquetGetForeignPlan(PlannerInfo *root,
                      RelOptInfo *baserel,
                      Oid foreigntableid,
                      ForeignPath *best_path,
                      List *tlist,
                      List *scan_clauses,
                      Plan *outer_plan);
extern TupleTableSlot *parquetIterateForeignScan(ForeignScanState *node);
extern void parquetBeginForeignScan(ForeignScanState *node, int eflags);
extern void parquetEndForeignScan(ForeignScanState *node);
extern void parquetReScanForeignScan(ForeignScanState *node);

PG_FUNCTION_INFO_V1(parquet_fdw_handler);
Datum
parquet_fdw_handler(PG_FUNCTION_ARGS)
{
	FdwRoutine *fdwroutine = makeNode(FdwRoutine);

	fdwroutine->GetForeignRelSize = parquetGetForeignRelSize;
	fdwroutine->GetForeignPaths = parquetGetForeignPaths;
	fdwroutine->GetForeignPlan = parquetGetForeignPlan;
	fdwroutine->BeginForeignScan = parquetBeginForeignScan;
	fdwroutine->IterateForeignScan = parquetIterateForeignScan;
	fdwroutine->ReScanForeignScan = parquetReScanForeignScan;
	fdwroutine->EndForeignScan = parquetEndForeignScan;

	PG_RETURN_POINTER(fdwroutine);
}

PG_FUNCTION_INFO_V1(parquet_fdw_validator);
Datum
parquet_fdw_validator(PG_FUNCTION_ARGS)
{
    List       *options_list = untransformRelOptions(PG_GETARG_DATUM(0));
    Oid         catalog = PG_GETARG_OID(1);
    ListCell   *lc;

    /* Only check table options */
    if (catalog != ForeignTableRelationId)
        PG_RETURN_VOID();

    foreach(lc, options_list)
    {
        DefElem    *def = (DefElem *) lfirst(lc);

        if (strcmp(def->defname, "filename") != 0
            && strcmp(def->defname, "sorted") != 0)
        {
            ereport(ERROR,
                    (errcode(ERRCODE_FDW_INVALID_OPTION_NAME),
                     errmsg("invalid option \"%s\"", def->defname)));
        }
    }

    PG_RETURN_VOID();
}

