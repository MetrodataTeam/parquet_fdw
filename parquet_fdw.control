# postgres_fdw extension
comment = 'foreign-data wrapper for parquet'
default_version = '0.1'
module_pathname = '$libdir/parquet_fdw'
relocatable = true
