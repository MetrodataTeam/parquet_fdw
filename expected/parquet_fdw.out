SET client_min_messages = WARNING;
CREATE EXTENSION parquet_fdw;
DROP ROLE IF EXISTS regress_parquet_fdw;
CREATE ROLE regress_parquet_fdw LOGIN SUPERUSER;
SET ROLE regress_parquet_fdw;
CREATE SERVER parquet_srv FOREIGN DATA WRAPPER parquet_fdw;
CREATE USER MAPPING FOR regress_parquet_fdw SERVER parquet_srv;
SET ROLE regress_parquet_fdw;
CREATE FOREIGN TABLE example (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename '/home/zilder/projects/parquet_fdw/data/example.parquet', sorted 'one');
SELECT * FROM example;
 one |    two     | three 
-----+------------+-------
   1 | {1,2,3}    | foo
   2 | {NULL,5,6} | bar
   3 | {7,8,9}    | baz
(3 rows)

-- sorting
EXPLAIN (COSTS OFF) SELECT * FROM example ORDER BY one;
       QUERY PLAN        
-------------------------
 Foreign Scan on example
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM example ORDER BY three;
          QUERY PLAN           
-------------------------------
 Sort
   Sort Key: three
   ->  Foreign Scan on example
(3 rows)

-- invalid options
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv;
ERROR:  parquet_fdw: filename is required
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename 'nonexistent.parquet', some_option '123');
ERROR:  parquet_fdw: No such file or directory
CREATE FOREIGN TABLE example2 (one INT8, two INT8[], three TEXT)
SERVER parquet_srv
OPTIONS (filename '/home/zilder/projects/parquet_fdw/data/example.parquet', some_option '123');
ERROR:  parquet_fdw: invalid option "some_option"
-- type mismatch
CREATE FOREIGN TABLE example2 (one INT8[], two INT8, three TEXT)
SERVER parquet_srv
OPTIONS (filename '/home/zilder/projects/parquet_fdw/data/example.parquet', sorted 'one');
SELECT one FROM example2;
ERROR:  parquet_fdw: incompatible types in column "one"
HINT:  parquet column is of scalar type while postgres type is array
SELECT two FROM example2;
ERROR:  parquet_fdw: incompatible types in column "two"
HINT:  parquet column is of type list while postgres type is scalar
